unic/char/property/src/macros.rs

```diff
@@ -108,62 +147,220 @@ macro_rules! char_property {
+            __impl StrTraitBinary for $name;
+            $bool_name;
+            $(#[$str_trait_meta])*; $str_trait_name;
+            $(#[$str_trait_fn_meta])*; $str_trait_fn_name;
+        }
+    };
+
+    // == Shared == //
+
+    ( __impl CharProperty for $name:ident;
+        $abbr:expr;
+        $long:expr;
+        $human:expr;
+    ) => {
+        impl $crate::CharProperty for $name {
+            fn prop_abbr_name() -> &'static str { $abbr }
+            fn prop_long_name() -> &'static str { $long }
+            fn prop_human_name() -> &'static str { $human }
+        }
+    };
+
+    ( __impl FromStr for $name:ident;
+        $( $id:ident => $value:expr; )*
+    ) => {
+        #[allow(unreachable_patterns)]
+        impl ::std::str::FromStr for $name {
+            type Err = ();
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                match s {
+                    // This stringify! should be moved out of this block to the call site. See the
+                    // test failure https://travis-ci.org/behnam/rust-unic/builds/275758001 for why
+                    // this is done here. This can be reverted at 1.20 adoption time.
+                    $( stringify!($id) => Ok($value), )*
+                    $( s if ::std::ascii::AsciiExt::eq_ignore_ascii_case(s, stringify!($id))
+                         => Ok($value), )*
+                    _ => Err(()),
                 }
             }
         }
+    };
 
+    ( __impl Display for $name:ident by $trait:ident ) => {
         impl ::std::fmt::Display for $name {
             fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
-                $crate::EnumeratedCharProperty::human_name(self).fmt(f)
+                $crate::$trait::human_name(self).fmt(f)
+            }
+        }
+    };
+
+    ( __impl CharTraitBinary for $name:ident;
+        $bool_name:ident;
+        $(#[$char_trait_meta:meta])*; $char_trait_name:ident;
+        $(#[$char_trait_fn_meta:meta])*; $char_trait_fn_name:ident;
+    ) => {
+        $(#[$char_trait_meta])*
+        pub trait $char_trait_name {
+            $(#[$char_trait_fn_meta])*
+            fn $char_trait_fn_name(self) -> bool;
+        }
+        impl $char_trait_name for char {
+            #[inline]
+            fn $char_trait_fn_name(self) -> bool {
+                $name::of(self).$bool_name()
+            }
+        }
+    };
+
+    ( __impl StrTraitBinary for $name:ident;
+        $bool_name:ident;
+        $(#[$str_trait_meta:meta])*; $str_trait_name:ident;
+        $(#[$str_trait_fn_meta:meta])*; $str_trait_fn_name:ident;
+    ) => {
+        $(#[$str_trait_meta])*
+        pub trait $str_trait_name {
+            $(#[$str_trait_fn_meta])*
+            fn $str_trait_fn_name(&self) -> bool;
+        }
+        impl $str_trait_name for str {
+            #[inline]
+            fn $str_trait_fn_name(&self) -> bool {
+                self.chars().any(|ch| $name::of(ch).$bool_name())
```

Comment:
I would have thought that a str impl would check `all`, not `any`.

Reply:
I can see where the has-a definition is convenient for the bidi properties. But unfortunately, I don't think that applies to all properties.